### ISSUE TITLE
feature request: add a per buffer bookmark supprt

### ISSUE THREAD

**Issue by xzbdmw** [otavioschwanck/arrow.nvim#15]
It will be cool to jump to buffer and then jump to bookmarks within this buffer in the same UI, mapped to a different toggle key, or just type ; and type any key to enter bookmark mode, since deleting, editing, and choosing bookmarks can have the same experience:)

**Comment by otavioschwanck** [comment]
> It will be cool to jump to buffer and then jump to bookmarks(narrowing scope to line) within this buffer in the same UI, mapped to a different toggle key, or just type ; and type any key to enter bookmark mode, since deleting, editing, and choosing bookmarks can have the same experience:) If you are interested I can try to make a pr!

It is like a marks, but just for the current buffer?
How could we show the lines on the arrow buffer?
Why not use just marks for this?

**Comment by otavioschwanck** [comment]
I Liked the idea, we could:

- Have a separate leader for buffer relative bookmarks
- We need the delete mode and the clear
- The indexes order should fallout the index_keys
- We can show the marked line dynamically (save only the line numbers for the cache)

**Comment by xzbdmw** [comment]
I was wandering If I can use tresitter to keep mark position even line number changed，because marks contantly changed during editibg is not that useful.

**Comment by otavioschwanck** [comment]
> I was wandering If I can use tresitter to keep mark position even line number changed，because marks contantly changed during editing is not that useful.

Nice idea! I will try to do it over the weekend (or the next one)

**Comment by otavioschwanck** [comment]
> I was wandering If I can use tresitter to keep mark position even line number changed，because marks contantly changed during editing is not that useful.

Hi, i tried to do with tree-sitter but i encountered a lot of challanges:

If the element is deleted, what should happen with the mark?
If the element is substituted, what should happen too?

Do you have any ideas of how to solve it? If you could do a simple POC (Prove of concept) with tree-sitter (Just a save / load position), i could implement it to arrow

**Comment by fnune** [comment]
My 2 cents (though I would personally not use a buffer-specific marks feature):

> If the element is deleted, what should happen with the mark?
> If the element is substituted, what should happen too?

In both cases the mark could be moved to the parent node?

**Comment by xzbdmw** [comment]
Yes either delete or move to parent according to user settings, through treesitter we can display symbol types such as function or classes displayed in ui, I now have time to implement a demo.

**Comment by xzbdmw** [comment]
I used extmark to track the bookmarks and recorded the context with navic. Would you be interested in integrating it into the arrow UI? Currently, persistence functionality is not implemented and only applies to a single buffer
```lua
return {
    "xzbdmw/bookmarktest",
    dependencies = {
        "SmiteshP/nvim-navic",
    },
    config = function()
        require("bookmarktest").setup()
    end,
}
```
use mm to add a bookmark and ml to bring up a list with context

**Comment by xzbdmw** [comment]
> > I was wandering If I can use tresitter to keep mark position even line number changed，because marks contantly changed during editing is not that useful.
> 
> Hi, i tried to do with tree-sitter but i encountered a lot of challanges:
> 
> If the element is deleted, what should happen with the mark? If the element is substituted, what should happen too?
> 
> Do you have any ideas of how to solve it? If you could do a simple POC (Prove of concept) with tree-sitter (Just a save / load position), i could implement it to arrow

for delete, add lines is tracked by extmarks, delete the marked line simply leave the mark with the orignal line, each time open the floating window I recaculate the linenumber to ensure correct jump

**Comment by otavioschwanck** [comment]
> ```lua
> {
>     "xzbdmw/bookmarktest",
>     dependencies = {
>         "SmiteshP/nvim-navic",
>     },
>     config = function()
>         require("bookmarktest").setup()
>     end,
> }
> ```

looks awesome.  This weekend i will bring it to arrow

**Comment by xzbdmw** [comment]
![image](https://github.com/otavioschwanck/arrow.nvim/assets/97848247/f15489ea-dfaf-43f8-8eb8-d124aeb08118)
I have successfully obtained the highlighted code for the bookmarks line and context, which is heavily borrowed from the treesitter context. I am thinking that displaying the entire context where the bookmarks line is located from beginning to end should be more intuitive(with other lines darker like the image above). Now, each bookmark uses a buffer and a window. What do you think?

**Comment by otavioschwanck** [comment]
> ![image](https://private-user-images.githubusercontent.com/97848247/313179808-f15489ea-dfaf-43f8-8eb8-d124aeb08118.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA1MDcxODIsIm5iZiI6MTcxMDUwNjg4MiwicGF0aCI6Ii85Nzg0ODI0Ny8zMTMxNzk4MDgtZjE1NDg5ZWEtZGZhZi00M2Y4LThlYjgtZDEyNGFlYjA4MTE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDEyNDgwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRmYWY4ZjY2OWJlZDZjZjZiMjJmZDA3OWY2OTVmYTY1MGRhZmQ1ZjliOGY2ZWJjODE1YjkxMjhmNGMwNzg2M2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.La23FSy6HSp0g8Fd_zegHFWNS4b4pCBr6DyeHtj-kXI) I have successfully obtained the highlighted code for the bookmarks line and context, which is heavily borrowed from treesitter-context. I am thinking that displaying the entire context where the bookmarks line is located from beginning to end should be more intuitive(with other lines darker like the image above,and use ... if there are too many lines). Now, each bookmark uses a buffer and a window. What do you think? The code is a mess and have some bugs now but you can try the basic functionality(mm to mark ml show list and cr to jump)

i loved the separated windows.  We could add the index at the title or something like that.

i already started doing the arrow code for the persist part.  At the weekend i will create a draft PR for it (and we can start looking for bugs / fixing)

**Comment by xzbdmw** [comment]
> > ![image](https://private-user-images.githubusercontent.com/97848247/313179808-f15489ea-dfaf-43f8-8eb8-d124aeb08118.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA1MDcxODIsIm5iZiI6MTcxMDUwNjg4MiwicGF0aCI6Ii85Nzg0ODI0Ny8zMTMxNzk4MDgtZjE1NDg5ZWEtZGZhZi00M2Y4LThlYjgtZDEyNGFlYjA4MTE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDEyNDgwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRmYWY4ZjY2OWJlZDZjZjZiMjJmZDA3OWY2OTVmYTY1MGRhZmQ1ZjliOGY2ZWJjODE1YjkxMjhmNGMwNzg2M2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.La23FSy6HSp0g8Fd_zegHFWNS4b4pCBr6DyeHtj-kXI) I have successfully obtained the highlighted code for the bookmarks line and context, which is heavily borrowed from treesitter-context. I am thinking that displaying the entire context where the bookmarks line is located from beginning to end should be more intuitive(with other lines darker like the image above,and use ... if there are too many lines). Now, each bookmark uses a buffer and a window. What do you think? The code is a mess and have some bugs now but you can try the basic functionality(mm to mark ml show list and cr to jump)
> 
> i loved the separated windows. We could add the index at the title or something like that.
> 
> i already started doing the arrow code for the persist part. At the weekend i will create a draft PR for it (and we can start looking for bugs / fixing)

That's great! Displaying a surrounding block is much simple and intuative compared to above approch
<img width="1207" alt="截屏2024-03-15 22 49 09" src="https://github.com/otavioschwanck/arrow.nvim/assets/97848247/89aaa88d-c8d2-4c3c-bf47-d9bd64982fdf">

**Comment by xzbdmw** [comment]
To achieve this effect, too much code and complexity have been added. Another way is to open a real buffer in the window, set the cursor line to the line recorded in the bookmark, and have an optional dependency on treesitter-context. If the user needs it, activate the context for each window, but the problem is that treesitter-context always keeps only one floating window on the currently active buffer

**Comment by cbochs** [comment]
@xzbdmw You might be interested in how [portal.nvim](https://github.com/cbochs/portal.nvim) accomplishes this 🙂 It could even be possible to tie leverage portal.nvim itself to do this for you?

**Comment by xzbdmw** [comment]
> @xzbdmw You might be interested in how [portal.nvim](https://github.com/cbochs/portal.nvim) accomplishes this 🙂 It could even be possible to leverage portal.nvim itself to do this for you?

Yes I tried it(I get the new idea there) but it lacks some ui to delete marks and swap order, possibly an action to expose some buffer marks globally, and some logic to persist marks per cwd.
By the way, it will be good to also have context support in https://github.com/cbochs/portal.nvim ,I'm looking to make a pr for that🙂

**Comment by cbochs** [comment]
> but it lacks some ui to delete marks and swap order, possibly an action to expose some buffer marks globally, and some logic to persist marks per cwd

That's what arrow is for 😉 

> Perhaps we can reuse the api to open preview and add edit mode and delete mode in arrow?

Hmm, it _might_ be possible to turn [this part](https://github.com/cbochs/portal.nvim/blob/main/lua/portal.lua#L56-L68) of Portal.tunnel into it's own API function (maybe, Portal.preview)? If that can be done then, then it should be relatively easy if arrow.nvim provides the search query and manages closing the selecting/closing the windows.

> By the way, it will be good to also have context support in https://github.com/cbochs/portal.nvim

That sounds like a great idea 👍

> I'm looking to make a pr for muti-window support 🙂

Nice!

**Comment by otavioschwanck** [comment]
@xzbdmw i had implemented all the persist part of this feature here:

https://github.com/otavioschwanck/arrow.nvim/pull/25/files

Is possible to save, remove by index, clear, all by the buffer.

Also the save / load data is all on json, so we can expand this part:  https://github.com/otavioschwanck/arrow.nvim/pull/25/files#diff-f6a5efa17b3949ec543c62aa762cedc52debd44b6376e3bd46d13be351618263R112 (Now is only saving the line)

From here we can decide what we need to persist and implement a way to show we wanna show the buffer marks.

PS: i will only have time at the weekend again, so if you want to implement the UI part from now on, be free! 

The Portal looks like a very nice idea

**Comment by xzbdmw** [comment]
> > but it lacks some ui to delete marks and swap order, possibly an action to expose some buffer marks globally, and some logic to persist marks per cwd
> 
> That's what arrow is for 😉
> 
> > Perhaps we can reuse the api to open preview and add edit mode and delete mode in arrow?
> 
> Hmm, it _might_ be possible to turn [this part](https://github.com/cbochs/portal.nvim/blob/main/lua/portal.lua#L56-L68) of Portal.tunnel into it's own API function (maybe, Portal.preview)? If that can be done then, then it should be relatively easy if arrow.nvim provides the search query and manages closing the selecting/closing the windows.
> 
> > By the way, it will be good to also have context support in https://github.com/cbochs/portal.nvim
> 
> That sounds like a great idea 👍
> 
> > I'm looking to make a pr for muti-window support 🙂
> 
> Nice!

Managed to implement the multi window and waiting for review, I’d like to try integrate portal and arrow, If something needs to change I’ll open a pr there👀
